lib/util.js

```diff
@@ -1057,3 +1057,68 @@ exports._exceptionWithHostPort = function(err,
 // process.versions needs a custom function as some values are lazy-evaluated.
 process.versions[exports.inspect.custom] =
   (depth) => exports.format(JSON.parse(JSON.stringify(process.versions)));
+
+const kCustomCallbackifyiedSymbol = Symbol('util.callbackify.custom');
+const kCustomCallbackifyiedToSplat = Symbol('util.callbackify.toSplat');
+const kCustomCallbackifyiedProject = Symbol('util.callbackify.project');
+
+function callbackify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to callbackify() ' +
+      'must be a function');
+  }
+
+  if (orig[kCustomCallbackifyiedSymbol]) {
+    const fn = orig[kCustomCallbackifyiedSymbol];
+    if (typeof fn !== 'function') {
+      throw new TypeError('The [util.callbackify.custom] property must be ' +
+        'a function');
+    }
+    Object.defineProperty(fn, kCustomCallbackifyiedSymbol, {
+      value: fn, enumerable: false, writable: false, configurable: true
+    });
+    return fn;
+  }
+
+  // we have two way to convert the returned value into multiple arguments
+  // either splat it or project some of it's keys
+  const toSplat = orig[kCustomCallbackifyiedToSplat];
+  const project = orig[kCustomCallbackifyiedProject];
+
+  const fn = function(...args) {
+    const cb = args.pop();
+    if (typeof cb !== 'function') {
+      throw new TypeError('The first argument to the callbackified function ' +
+        'must be a callback style function');
+    }
+    orig.call(this, ...args).then((ret) => {
+      if (toSplat) {
+        cb(null, ...ret);
+      } else if (project) {
+        const allThere = project.all((k) => k in ret);
+        if (!allThere) {
+          throw new TypeError('Not all of the projected properties' +
+            'are present in the returned value: ' +
+            inspect(ret));
+        }
+        const args = project.reduce((s, k) => s.concat(ret[k]), []);
+        cb(null, ...args);
+      } else {
+        cb(null, ret);
+      }
+    }, cb);
```

Comment:
This could cause confusing behavior if the original Promise rejects with a non-error:

```js
function promiseFunction() {
  return Promise.reject(null);
}

util.callbackify(promiseFunction)((err, result) => {
  // err will be null, which could be confusing because the promise actually rejected.
});
```

Bluebird seems to handle this by [wrapping the rejection reason in an `Error` if it's falsy](https://github.com/petkaantonov/bluebird/blob/2207fae3572f03b089bc92d3a6cefdd278cff7ab/src/nodeify.js#L30-L43).